### PR TITLE
Added dot function

### PR DESCRIPTION
**Pull request by martingochevski**
- added dot function to jax, numpy, tensorflow and torch backends as well as Ivy with appropriate array and container instance methods

### CODE REVIEW

**Comment by jkeane508** [comment]
> * added dot function to jax, numpy, tensorflow and torch backends as well as Ivy with appropriate array and container instance methods

Hello,

Can you please link the issue that this PR closes?

Thanks!

**Comment by martingochevski** [comment]
Closes #16656

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 1]
```diff
@@ -133,6 +134,16 @@ def kron(
         """
         return ivy.kron(self._data, b, out=out)
 
+    def dot(
+        self: ivy.Array,
+        b: ivy.Array,
+        /,
+        *,
+        out: Optional[ivy.Array] = None,
+        # precision: Optional[jax.lax.Precision] = None
```
Comment (jkeane508) [inline ivy/data_classes/array/experimental/linear_algebra.py:143]:
```suggestion
```

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 2]
```diff
@@ -780,6 +782,97 @@ def multi_dot(
...
+        x1: Union[ivy.Array, ivy.NativeArray, ivy.Container],
+        x2: Union[ivy.Container, ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        key_chains: Optional[Union[List[str], Dict[str, str]]] = None,
+        to_apply: bool = True,
+        prune_unapplied: bool = False,
+        map_sequences: bool = False,
+        out: Optional[ivy.Container] = None,
+        precision: Optional[jax.lax.Precision] = None
```
Comment (jkeane508) [inline ivy/data_classes/container/experimental/linear_algebra.py:797]:
```suggestion
```

We shouldn't have any Jax specific objects in the Ivy API

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
 # global
 from typing import Union, Optional, List, Dict, Tuple, Sequence
 
+import jax
+
 # local
 from ivy.data_classes.container.base import ContainerBase
 import ivy
```
Comment (jkeane508) [inline ivy/data_classes/container/experimental/linear_algebra.py:8]:
```suggestion
```

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 4]
```diff
@@ -780,6 +782,97 @@ def multi_dot(
...
+        >>> a = ivy.Container(x=ivy.arange(6),
+        ...                   y=ivy.arange(6))
+        >>> ivy.Container.static_dot((a))
+        """
+        return ContainerBase.cont_multi_map_in_function(
+            "dot",
+            x1,
+            x2,
+            out=out,
+            # precision=precision,
```
Comment (jkeane508) [inline ivy/data_classes/container/experimental/linear_algebra.py:835]:
```suggestion
```

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 5]
```diff
@@ -780,6 +782,97 @@ def multi_dot(
...
+        self: ivy.Container,
+        x2: Union[ivy.Container, ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        key_chains: Optional[Union[List[str], Dict[str, str]]] = None,
+        to_apply: bool = True,
+        prune_unapplied: bool = False,
+        map_sequences: bool = True,
+        out: Optional[ivy.Container] = None,
+        # precision: Optional[jax.lax.Precision] = None
```
Comment (jkeane508) [inline ivy/data_classes/container/experimental/linear_algebra.py:852]:
```suggestion
```

File: ivy/data_classes/container/experimental/linear_algebra.py [hunk 6]
```diff
@@ -780,6 +782,97 @@ def multi_dot(
...
+        """
+        return self.static_dot(
+            self,
+            x2,
+            key_chains=key_chains,
+            to_apply=to_apply,
+            prune_unapplied=prune_unapplied,
+            map_sequences=map_sequences,
+            out=out,
+            # precision=precision
```
Comment (jkeane508) [inline ivy/data_classes/container/experimental/linear_algebra.py:873]:
```suggestion
```

File: ivy/functional/backends/jax/experimental/linear_algebra.py [hunk 7]
```diff
@@ -140,6 +140,19 @@ def multi_dot(
     return jnp.linalg.multi_dot(x)
 
 
+def dot(
+    x1: JaxArray,
+    x2: JaxArray,
+    out: Optional[JaxArray] = None,
+    precision: Optional[Precision] = None
```
Comment (jkeane508) [inline ivy/functional/backends/jax/experimental/linear_algebra.py:147]:
```suggestion
```

File: ivy/functional/backends/tensorflow/experimental/linear_algebra.py [hunk 8]
```diff
@@ -1,4 +1,6 @@
 from typing import Union, Optional, Tuple, List, Sequence
+
+import numpy as np
```
Comment (jkeane508) [inline ivy/functional/backends/tensorflow/experimental/linear_algebra.py:3]:
```suggestion
```

File: ivy/functional/backends/tensorflow/experimental/linear_algebra.py [hunk 9]
```diff
@@ -150,6 +152,18 @@ def multi_dot(
...
+def dot(
+    x1: tf.Tensor,
+    x2: tf.Tensor,
+    out: Optional[tf.Tensor] = None
+) -> tf.Tensor:
+    x1, x2 = ivy.promote_types_of_inputs(x1, x2)
+    return tf.experimental.numpy.dot(x1, x2)
+
+
+# dot.support_native_out = True
```
Comment (jkeane508) [inline ivy/functional/backends/tensorflow/experimental/linear_algebra.py:164]:
```suggestion
```

File: ivy/functional/backends/tensorflow/experimental/linear_algebra.py [hunk 10]
```diff
@@ -150,6 +152,18 @@ def multi_dot(
     return dot_out
 
 
+def dot(
+    x1: tf.Tensor,
+    x2: tf.Tensor,
+    out: Optional[tf.Tensor] = None
+) -> tf.Tensor:
+    x1, x2 = ivy.promote_types_of_inputs(x1, x2)
+    return tf.experimental.numpy.dot(x1, x2)
```
Comment (jkeane508) [inline ivy/functional/backends/tensorflow/experimental/linear_algebra.py:161]: This should probably be [`tensordot()`](https://www.tensorflow.org/api_docs/python/tf/tensordot)

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 11]
```diff
@@ -1,6 +1,8 @@
 # global
 from typing import Union, Optional, Tuple, List, Sequence
 
+import jax.lax
```
Comment (jkeane508) [inline ivy/functional/ivy/experimental/linear_algebra.py:4]:
```suggestion
```

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 12]
```diff
@@ -502,6 +504,62 @@ def multi_dot(
...
+@handle_out_argument
+@to_native_arrays_and_back
+@handle_exceptions
+def dot(
+    a: Union[ivy.Array, ivy.NativeArray],
+    b: Union[ivy.Array, ivy.NativeArray],
+    /,
+    *,
+    out: Optional[Union[ivy.Array, ivy.NativeArray]] = None,
+    precision: Optional[jax.lax.Precision] = None
```
Comment (jkeane508) [inline ivy/functional/ivy/experimental/linear_algebra.py:517]:
```suggestion
```

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 13]
```diff
@@ -502,6 +504,62 @@ def multi_dot(
...
+    # >>> C = ivy.zeros(100)
+    # >>> ivy.dot(A, B, out=C)
+    # >>> print(C)
+    #
+    # >>> A = ivy.arange(100)
+    # >>> B = ivy.arange(100)
+    # >>> ivy.dot(A, B, precision=jax.lax.Precision('high'))
+
+    """
+    return current_backend(a, b).dot(a, b, out=out, precision=precision)
```
Comment (jkeane508) [inline ivy/functional/ivy/experimental/linear_algebra.py:560]:
```suggestion
    return current_backend(a, b).dot(a, b, out=out,)
```

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 14]
```diff
@@ -502,6 +504,62 @@ def multi_dot(
...
+    #
+    # >>> A = ivy.arange(100)
+    # >>> B = ivy.arange(100)
+    # >>> C = ivy.zeros(100)
+    # >>> ivy.dot(A, B, out=C)
+    # >>> print(C)
+    #
+    # >>> A = ivy.arange(100)
+    # >>> B = ivy.arange(100)
+    # >>> ivy.dot(A, B, precision=jax.lax.Precision('high'))
```
Comment (jkeane508) [inline ivy/functional/ivy/experimental/linear_algebra.py:557]:
```suggestion
    # >>> ivy.dot(A, B,)
```

File: ivy/functional/backends/jax/experimental/linear_algebra.py [hunk 15]
```diff
@@ -140,6 +140,19 @@ def multi_dot(
...
+) -> JaxArray:
+    x1, x2 = ivy.promote_types_of_inputs(x1, x2)
+    return jnp.dot(a=x1, b=x2, out=out, precision=precision)
```
Comment (jkeane508) [inline ivy/functional/backends/jax/experimental/linear_algebra.py:150]:
```suggestion
    return jnp.dot(a=x1, b=x2, out=out,)
```

File: ivy/functional/backends/jax/experimental/linear_algebra.py [hunk 16]
```diff
@@ -3,7 +3,7 @@
 import jax.numpy as jnp
 import jax.scipy.linalg as jla
 from ivy.functional.backends.jax import JaxArray
-
+from jax.lax import Precision
```
Comment (jkeane508) [inline ivy/functional/backends/jax/experimental/linear_algebra.py:6]:
```suggestion
```

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 17]
```diff
@@ -1,6 +1,7 @@
 # global
 import abc
 from typing import Optional, Union, Tuple, List, Sequence
+import jax.lax
```
Comment (jkeane508) [inline ivy/data_classes/array/experimental/linear_algebra.py:4]:
```suggestion
```

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Hey,

Just need to remove the precision argument from this. Looks like this is JAX specific behaviour and therefore we should just not include it. We are aiming to create an abstraction over all the frameworks, so any framework specific objects can be ignored. (We can handle this behaviour directly in the frontends)

Feel free to ask any questions!

**Comment by jkeane508** [comment]
ivy-gardener

File: ivy/functional/backends/tensorflow/experimental/linear_algebra.py (hunk 10 above)
Comment (martingochevski) [reply ivy/functional/backends/tensorflow/experimental/linear_algebra.py:156]: I went with the numpy dot function, since tf.tensordot adds overhead in attempting to handle the axes which seems irrelevant here as Ivy already has a full fledged [tensordot function](https://github.com/unifyai/ivy/blob/master/ivy/functional/ivy/linear_algebra.py#L2113).

File: ivy/functional/backends/tensorflow/experimental/linear_algebra.py (hunk 9 above)
Comment (jkeane508) [reply ivy/functional/backends/tensorflow/experimental/linear_algebra.py:159]: Just need to remove this!

**Review by jkeane508 — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/tensorflow/experimental/linear_algebra.py (hunk 9 above)
Comment (martingochevski) [reply ivy/functional/backends/tensorflow/experimental/linear_algebra.py]: Done and dusted.

**Comment by jkeane508** [comment]
ivy-gardener

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by jkeane508** [comment]
> Personally I wouldn’t call it stale.

Hey,

It's just automation for PRs that haven't had any activity for 2 weeks. Will remove the label :)

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Hey,

Seems that there are tests for this failing [here](https://github.com/unifyai/ivy/actions/runs/5463170502/jobs/9943519355?pr=16978#step:3:1307). We would need to get these passing before we can merge :)

Thanks!

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 18]
```diff
@@ -490,6 +490,60 @@ def test_multi_dot(dtype_x, test_flags, backend_fw, fn_name):
...
+    dtype_xy=helpers.dtype_and_values(
+        available_dtypes=(ivy.float32, ivy.float64),
+        num_arrays=2,
+        large_abs_safety_factor=8,
+        small_abs_safety_factor=8,
+        safety_factor_scale="log",
+        min_num_dims=1,
+        max_num_dims=1,
+    ),
+    ground_truth_backend="numpy",
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:504]: I think this is throwing an error

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Merged your changes so that tests could run. But it looks like tests are failing due to a handle in the test

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 18 above)
Comment (martingochevski) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:504]: Ah, I thought it made sense to use numpy as the ground truth since most of the backend implementations rely on it in the end.

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.
